File: library/base/src/main/java/org/cru/godtools/base/util/LocaleUtils.java
package org.cru.godtools.base.util;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;

import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.os.ConfigurationCompat;
import timber.log.Timber;

import static org.ccci.gto.android.common.util.LocaleUtils.getOptionalDisplayName;

public class LocaleUtils {
    private static final String STRING_RES_LANGUAGE_NAME_PREFIX = "language_name_";

    public static Locale getDeviceLocale(@NonNull final Context context) {
        return ConfigurationCompat.getLocales(context.getResources().getConfiguration()).get(0);
    }

    @NonNull
    public static String getDisplayName(@NonNull final Locale locale, @Nullable Context context,
                                        @Nullable final String defaultName, @Nullable final Locale inLocale) {
        // check for a language name string resource
        if (context != null) {
            if (inLocale != null) {
                context = localizeContextIfPossible(context, inLocale);
            }
            final String name = getLanguageNameStringRes(context, locale);
            if (name != null) {
                return name;
            }
        }

        // use Locale.getDisplayName()
        final String name = getOptionalDisplayName(locale, inLocale);
        if (name != null) {
            return name;
        }

        // use the default name if specified
        if (defaultName != null) {
            return defaultName;
        }

        // just rely on Locale.getDisplayName() which will default to the language code at this point
        Timber.tag("LocaleUtils")
                .e(new RuntimeException("Unable to find display name for " + locale.toString()),
                   "LocaleUtils.getDisplayName(%s, %s)", locale, inLocale);
        return inLocale != null ? locale.getDisplayName(inLocale) : locale.getDisplayName();
    }

    @NonNull
    public static Context localizeContextIfPossible(@NonNull final Context context, @NonNull final Locale locale) {
        final Configuration conf = new Configuration(context.getResources().getConfiguration());
        conf.setLocale(locale);
        return context.createConfigurationContext(conf);
    }

    @Nullable
    private static String getLanguageNameStringRes(@NonNull final Context context, @NonNull final Locale locale) {
        final Resources resources = context.getResources();
        final int stringId = resources
                .getIdentifier(STRING_RES_LANGUAGE_NAME_PREFIX + locale.toString().toLowerCase(Locale.ENGLISH),
                               "string", context.getPackageName());
        if (stringId == 0) {
            return null;
        }

        return resources.getString(stringId);
    }
}

File: library/base/src/main/java/org/cru/godtools/base/util/LocaleUtils.kt
@file:JvmName("LocaleUtils")

package org.cru.godtools.base.util

import android.content.Context
import androidx.annotation.VisibleForTesting
import androidx.core.os.ConfigurationCompat
import java.util.Locale
import org.ccci.gto.android.common.util.content.localize
import org.ccci.gto.android.common.util.getOptionalDisplayName
import timber.log.Timber

@VisibleForTesting
internal const val STRING_RES_LANGUAGE_NAME_PREFIX = "language_name_"

val Context.deviceLocale: Locale get() = ConfigurationCompat.getLocales(resources.configuration)[0]

@JvmOverloads
fun Locale.getDisplayName(context: Context? = null, defaultName: String? = null, inLocale: Locale? = null): String {
    return context?.localizeIfPossible(inLocale)?.getLanguageNameStringRes(this)
        // use Locale.getDisplayName()
        ?: getOptionalDisplayName(inLocale)
        // use the default name if specified
        ?: defaultName
        // just rely on Locale.getDisplayName() which will default to the language code at this point
        ?: run {
            Timber.tag("LocaleUtils").e(
                RuntimeException("Unable to find display name for $this"),
                "LocaleUtils.getDisplayName(%s, %s)", this, inLocale
            )
            if (inLocale != null) getDisplayName(inLocale) else displayName
        }
}

fun Context.localizeIfPossible(locale: Locale?) = locale?.let { localize(it) } ?: this

private fun Context.getLanguageNameStringRes(locale: Locale) =
    when (val stringId = resources.getIdentifier(locale.languageNameStringRes, "string", packageName)) {
        0 -> null
        else -> resources.getString(stringId)
    }

private val Locale.languageNameStringRes
    get() = "$STRING_RES_LANGUAGE_NAME_PREFIX${toString().toLowerCase(Locale.ENGLISH)}"
